group-coordinator/src/main/java/org/apache/kafka/coordinator/group/OffsetMetadataManager.java

```diff
@@ -0,0 +1,454 @@
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.OFFSET_METADATA_TOO_LARGE.code()));
+                } else {
+                    log.debug("[GroupId {}] Committing offsets {} for partition {}-{} from member {} with leader epoch {}.",
+                        request.groupId(), partition.committedOffset(), topic.name(), partition.partitionIndex(),
+                        request.memberId(), partition.committedLeaderEpoch());
+
+                    topicResponse.partitions().add(new OffsetCommitResponseData.OffsetCommitResponsePartition()
+                        .setPartitionIndex(partition.partitionIndex())
+                        .setErrorCode(Errors.NONE.code()));
+
+                    final OptionalInt leaderEpoch = partition.committedLeaderEpoch() == RecordBatch.NO_PARTITION_LEADER_EPOCH ?
+                        OptionalInt.empty() : OptionalInt.of(partition.committedLeaderEpoch());
+                    final String metadata = partition.committedMetadata() == null ?
+                        OffsetAndMetadata.NO_METADATA : partition.committedMetadata();
+                    final long commitTimestampMs = partition.commitTimestamp() == OffsetCommitRequest.DEFAULT_TIMESTAMP ?
+                        currentTimeMs : partition.commitTimestamp();
+
+                    final OffsetAndMetadata offsetAndMetadata = new OffsetAndMetadata(
+                        partition.committedOffset(),
+                        leaderEpoch,
+                        metadata,
+                        commitTimestampMs,
+                        expireTimestampMs
+                    );
+
+                    records.add(RecordHelpers.newOffsetCommitRecord(
+                        request.groupId(),
+                        topic.name(),
+                        partition.partitionIndex(),
+                        offsetAndMetadata,
+                        metadataImage.features().metadataVersion()
+                    ));
+                }
+            });
+        });
+
+        return new CoordinatorResult<>(records, response);
+    }
+
+    /**
+     * Replays OffsetCommitKey/Value to update or delete the corresponding offsets.
+     *
+     * @param key   A OffsetCommitKey key.
+     * @param value A OffsetCommitValue value.
+     */
+    public void replay(
+        OffsetCommitKey key,
+        OffsetCommitValue value
+    ) {
+        final String groupId = key.group();
+        final TopicPartition tp = new TopicPartition(key.topic(), key.partition());
+
+        if (value != null) {
+            // Ensures that there is a corresponding group for the offsets. If a group does
+            // not exist, a generic group is created as a so-called "simple group".
+            try {
+                groupMetadataManager.group(groupId);
+            } catch (GroupIdNotFoundException ex) {
+                groupMetadataManager.getOrMaybeCreateGenericGroup(groupId, true);
+            }
```

Comment:
this looks to be for the loading scenario. don't we need to create a new group for consumer groups too? i'm not sure why the generic group gets special treatment.

Reply:
How would you do it based on the offset record? We don't know if the offset is for a generic group or a consumer group.

Note that the group should already exists when offsets are committed. There is only one case when it does not. It is in the "simple consumer group" case where Kafka is used for storing offsets but not for doing the membership. Those groups are empty by definition but contains offsets. Unfortunately, we don't have group metadata records for those in the current implementation (in scala) so we have to ensure that a group exists here.

Reply:
that was my line of thinking as well. i would think that we would always have a group already created before replaying an offset commit record.

so if the simple group is an edge case we're handling for generic groups, why don't we need to handle it for consumer groups?

Reply:
We don't have this case for (new) consumer groups as they can only be used with subscriptions.

For the context, I think that we could say that we have three types of groups: a) simple consumer group, b) old consumer (or new  generic) group, and c) new consumer group. a) is basically an old consumer group with no members and no protocol type. This is why we create a generic group in this case.

I was also considering whether we should introduce this group as a new type in our new architecture. The issue is that it is possible to migrate from a) to b). By reusing a generic group, this is automatic. Otherwise, we would have to handle to transition. I think that we could still consider this but I would decouple this from this PR. What's your take on this?

Reply:
gotcha. technically a simple group is a subset of a generic group right? not sure if it should have its own type

Reply:
Yeah, I think for most purposes the simple group is a generic group and making a new type might be confusing if a can migrate to b.